sentry-core/src/main/java/io/sentry/core/SentryClient.java

```diff
@@ -88,43 +88,83 @@ public SentryClient(final @NotNull SentryOptions options, @Nullable Connection c
                 SentryLevel.DEBUG,
                 "Event was dropped by processor: %s",
                 processor.getClass().getName());
+        sendEvent = false;
         break;
       }
     }
 
-    if (event == null) {
-      return SentryId.EMPTY_ID;
-    }
+    Session session = null;
 
-    // TODO: should it be done on the HUB?
-    updateSessionData(event, hint, scope);
+    if (event != null) {
+      session = updateSessionData(event, hint, scope);
 
-    if (!sample()) {
-      options
-          .getLogger()
-          .log(
-              SentryLevel.DEBUG,
-              "Event %s was dropped due to sampling decision.",
-              event.getEventId());
-      return SentryId.EMPTY_ID;
+      if (!sample()) {
+        options
+            .getLogger()
+            .log(
+                SentryLevel.DEBUG,
+                "Event %s was dropped due to sampling decision.",
+                event.getEventId());
+        sendEvent = false;
+        event = null;
+      }
     }
 
-    event = executeBeforeSend(event, hint);
+    if (event != null) {
```

Comment:
I find that the combination of the boolean `sendEvent` and the multiple null check is too confusing. I found it hard to follow to make sure this should work in all cases.

What about we just guard at the start for null (and return), then at the two locations where user callbacks can set it to null?

Reply:
i will look to get rid of one of them, but i had a reason to not do so, just dont remember now, tomorrow i look into it